File: flask_app/app/services/address_book_service.py
from app.dao.address_book_dao import AddressBookDAO
from app.utils.exceptions import NotFoundException

class AddressBookService:
    @staticmethod
    def create_address(user_id, data):
        data['user_id'] = user_id
        return AddressBookDAO.create(data)

    @staticmethod
    def get_addresses(user_id):
        return AddressBookDAO.get_by_user(user_id)
    
    @staticmethod
    def get_addresses_by_id(address_id):
        return AddressBookDAO.get_by_id(address_id)

    @staticmethod
    def update_address(address_id, user_id, updates):
        address = AddressBookDAO.get_by_id(address_id)
        if not address or address.user_id != user_id:
            raise NotFoundException("Address not found or unauthorized")
        return AddressBookDAO.update(address, updates)

    @staticmethod
    def delete_address(address_id, user_id):
        address = AddressBookDAO.get_by_id(address_id)
        if not address or address.user_id != user_id:
            raise NotFoundException("Address not found or unauthorized")
        AddressBookDAO.delete(address)

File: springboot_api/src/main/java/com/example/membership_api/service/AddressBookService.java
package com.example.membership_api.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.membership_api.dao.AddressBookDAO;
import com.example.membership_api.exception.NotFoundException;
import com.example.membership_api.model.AddressBook;

@Service
public class AddressBookService {

    @Autowired
    private AddressBookDAO addressBookDAO;

    public AddressBook createAddress(String userId, AddressBook data) {
        data.setUserId(userId);
        return addressBookDAO.create(data);
    }

    public List<AddressBook> getAddresses(String userId) {
        return addressBookDAO.getByUser(userId);
    }

    public AddressBook getAddressById(String addressId) {
        AddressBook address = addressBookDAO.getById(addressId);
        if (address == null) {
            throw new NotFoundException("Address not found or unauthorized");
        }
        return address;
    }

    public AddressBook updateAddress(String addressId, String userId, AddressBook updates) {
        AddressBook address =  addressBookDAO.getByIdUserId(addressId, userId);
        return addressBookDAO.update(address, updates);
    }

    public void deleteAddress(String addressId, String userId) {
        AddressBook address = addressBookDAO.getByIdUserId(addressId, userId);
        addressBookDAO.delete(address);
    }
}
